File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / (orbitalFrac planet * baseOrbitalSeconds)

baseOrbitalSeconds :: Float
baseOrbitalSeconds = 31557600

orbitalFrac :: Planet -> Float
orbitalFrac Earth = 1
orbitalFrac Mercury = 0.2408467
orbitalFrac Venus = 0.61519726
orbitalFrac Mars = 1.8808158
orbitalFrac Jupiter = 11.862615
orbitalFrac Saturn = 29.447498
orbitalFrac Uranus = 84.016846
orbitalFrac Neptune = 164.79132

File: nim/space-age/space_age.nim
type
  Planet* = enum
    Mercury, Venus, Earth, Mars, Jupiter, Saturn, Uranus, Neptune

proc age*(planet: Planet, seconds: int64): float =
  seconds.float / 31557600 / (case planet
                              of Mercury: 0.2408467
                              of Venus: 0.61519726
                              of Earth: 1
                              of Mars: 1.8808158
                              of Jupiter: 11.862615
                              of Saturn: 29.447498
                              of Uranus: 84.016846
                              of Neptune: 164.79132)
  
